desc/objectives/_stability.py

```diff
@@ -0,0 +1,283 @@
+import numpy as np
+from desc.utils import Timer
+from desc.grid import LinearGrid
+from desc.transform import Transform
+from desc.compute import data_index, compute_mercier_stability, compute_magnetic_well
+from .objective_funs import _Objective
+
+
+class MercierStability(_Objective):
+    """The Mercier criterion is a fast proxy for MHD stability.
+
+    This makes it a useful figure of merit for stellarator operation.
+    Systems with DMerc > 0 are favorable for stability.
+
+    See equation 4.16 in
+    Landreman, M., & Jorge, R. (2020). Magnetic well and Mercier stability of
+    stellarators near the magnetic axis. Journal of Plasma Physics, 86(5), 905860510.
+    doi:10.1017/S002237782000121X.
+
+    Parameters
+    ----------
+    eq : Equilibrium, optional
+        Equilibrium that will be optimized to satisfy the Objective.
+    target : float, ndarray, optional
+        Target value(s) of the objective.
+        len(target) must be equal to Objective.dim_f
+    weight : float, ndarray, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        len(weight) must be equal to Objective.dim_f
+    grid : LinearGrid, ConcentricGrid, QuadratureGrid, ndarray, optional
+        Collocation grid containing the nodes to evaluate at.
+    name : str
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _linear = False
+
+    def __init__(
+        self, eq=None, target=0, weight=1, grid=None, name="Mercier Stability"
+    ):
+        self.grid = grid
+        super().__init__(eq=eq, target=target, weight=weight, name=name)
+        self._callback_fmt = "Mercier Stability: {:10.3e}"
+
+    def build(self, eq, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        eq : Equilibrium, optional
+            Equilibrium that will be optimized to satisfy the Objective.
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        if self.grid is None:
+            self.grid = LinearGrid(
+                M=eq.M_grid,
+                N=eq.N_grid,
+                NFP=eq.NFP,
+                sym=eq.sym,
+                rho=np.linspace(1 / 5, 1, 5),
```

Comment:
I should change this to` jnp.linspace`

Reply:
I purposely changed this to use `numpy` instead of `jax.numpy`. All of the grid, basis, and transform objects are pre-computed and held constant during optimization. JAX arrays are only needed for quantities that are being differentiated with respect to. It shouldn't matter in practice, but we have been using regular NumPy arrays for anything that we know doesn't need JAX